### ISSUE TITLE
Best answer

### ISSUE THREAD

**Issue by ludwiczakpawel** [forem/forem#16306]
### Description

Admins want an ability to have "Q&A" type of posts where it would be possible to mark post as answered with specific comment.

There are tons of directions we could pull this towards but lets kick off with simple MVP: lets just enable post authors to mark specific comment as “Best Answer” - it would get highlighted with special styling and it would land on top of the Discussion list.

For sake of simplicity let's limit this to only top level comments (in other words: it won't be possible to mark comment's reply as "best answer")

### Design

![image](https://user-images.githubusercontent.com/108287/151126420-8118a1df-61cd-4041-8c40-663867a2d226.png)

![image](https://user-images.githubusercontent.com/108287/151126446-b9c316b0-eba9-4873-bec2-5a9dd4bddb2c.png)

[Figma](https://www.figma.com/file/Rb134nTqUACqyTJnPTGg94/The-Best-Answer?node-id=0%3A1).

### TBD

- Who should be able to mark comments as best answers?
    - Post authors - for sure
    - Admins - i guess so..
    - Mods - ?

### Future

Some ideas for future, just to keep in mind but definitely **not to implement now**:

1. adding an indicator somewhere on top of the post view saying that this post has “best answer” chosen so user doesn’t have to scroll down to know it.
2. adding similar indicator in the Feed (maybe as an icon in front of the title or something, whatever)
3. potentially, when a post has “best answer” chosen we could automatically [show that answer in the comments section right in the feed](https://d.pr/i/CcwkpI)..

**Comment by ludwiczakpawel** [comment]
@aitchiss I'm not language expert sooo I leave the decision to others.. But since you brought it up there're two more I'm thinking about:
- "Highlighted comment"
- "Pinned comment" - as reference to pinned posts we have in feed..

**Comment by amywtlin** [comment]
@JennieOcken Clarifying a bit more on the goal for "best answer" here before I weigh in on how we should describe a picked comment. 

Is the main goal here to...
1. make a comment more prominent (leaning towards pinning a comment in a neutral way, not necessarily "resolved")
2. make an unresolved post as resolved (this implies a closure to the post, which calls for different communications than the former)

I am supposed we are focusing more on #2, but is that correct?

**Comment by ellativity** [comment]
Ooooh, just to make things more complicated, could we have a "mark as resolved" check box with an option that allows the marker (assuming this is post author but admins could modify as with most things) to select the "most helpful" answer?

**Comment by ellativity** [comment]
Sorry! I clicked on the wrong button - hope I didn't mess things up for you @ludwiczakpawel

**Comment by JennieOcken** [comment]
I think there is a ton of value in the "pin comment" option even without the workflow of "mark as resolved".  Given that, let's take a more iterative approach.  For this issue, let's focus on a more "pin comment" or "highlight comment" option.  That will give the Creators I have heard from the majority of what they are looking for.  Then we can let it rest and see what kind of response we get.  In the future, we may add some of the "resolved" type concepts, but that can be a succinct additional project.   Also, I don't see the "pin comment" approach precluding a later "resolved" approach.   Make sense?

**Comment by amywtlin** [comment]
> For this issue, let's focus on a more "pin comment" or "highlight comment" option. That will give the Creators I have heard from the majority of what they are looking for.

Thanks - this is what I was looking for. Yes, we should take an iterative approach, and I would love to make sure what we work on first is addressing the majority of what Creators are looking for.

@aitchiss @ludwiczakpawel Let's focus on "making a comment more prominent" then. Going back to the wording, for this use case, best answer won't be a suitable name. I don't have a strong opinion here, as both "pinning" or "highlighting" a post feel neutral to me and can be applicable in lots of scenarios, no matter for what reasons an author or an admin decides to make a comment more prominent. Do you two have a preference?

Question: can there be more than one post that are pinned though? No?

**Comment by aitchiss** [comment]
> I don't have a strong opinion here, as both "pinning" or "highlighting" a post feel neutral to me and can be applicable in lots of scenarios, no matter for what reasons an author or an admin decides to make a comment more prominent. Do you two have a preference?

Yeah I agree, I think either works well. Like @ludwiczakpawel mentioned "pinned comment" is consistent with the language we use for pinning posts, so maybe all other things being equal, we should go for that?

> Question: can there be more than one post that are pinned though? No?

I'd assumed we were just going for 1 pinned comment, but it's a good question! Even if we stick with just 1 for the first iteration, it would be good to know if we envisage this expanding to multiple pinned comments, just to make sure the implementation allows for that in future.

**Comment by ludwiczakpawel** [comment]
> could we have a "mark as resolved" check box

@ellativity What do you envision this would do? In other words: how would that be different from "pinning comment" (well, except language)?

---

I don't want to go too granular here and have too many "states" for a comment.

---

> Question: can there be more than one post that are pinned though? 

I'd say "no" for now to not overcomplicate things.

Correct me if I'm wrong, @JennieOcken , but I think this feature was requested by Creators who want to have some sort of Q&A type of posts.. With that in mind, lets try to focus on that use case.

Usually Q&A sites (i.e. Stack Overflow 😭 ) will let you mark only one answer as "correct". So lets start with that.
If at some point we feel like "Marking as resolved" (per @ellativity 's suggestion) is actually the same thing as pinning/highlighting a comment, then it would be easier to transition. The same if at any given point we see a need for Creators to mark multiple comments as highlighted/pinned – we can easily scale it up. But it won't be that easy to go the other way (i.e. if we kick off by letting Creators mark multiple comments as highlighted/pinned and then we realize we don't need that and we only want one comment to be marked).

So my suggestion: let's go with only being able to mark one comment. Rest is, I think, matter of naming ("best answer" vs "highlighted" vs ...) and that should be based on what Creators really want (cc @JennieOcken ).

**Comment by JennieOcken** [comment]
The original request was absolutely some sort of Q&A type posts.  But we have also heard similar requests over time for cases like CodeNewbie or Lee's Forems where you want to highlight specific comments.

I agree that we should keep it to the concept of "pinning" a comment (liking that wording since it is already in the ubiquitous language of the site) since it is addressing the main request (Q&A) but not boxing us into a corner.  This may be not _exactly_ how Creators talk about it, but it matches the nature of our application.

Also agree that we should do only one for now.  Especially since we limit pinned posts to only one.  And there becomes some interesting logic on which order you list pinned items in which is it's own box I don't want to open right now.

**Comment by ellativity** [comment]
I see, exactly as @JennieOcken has described, 2 clear leading use cases for this: 

1. Q&A type posts, where it might be nice to have a "Resolved" flag or tag at some point. The value of the tag would be that all resolved answers could be found on a tag page, or if a user is searching for a question they can see at a glance which of the results contain questions that have been resolved instead of clicking into all of them to check if there's a pinned comment.
2. [this post on forem.dev](https://forem.dev/foremteam/changelog-contact-email-and-smtp-settings-customization-2kpc) where Lee added a comment that helped highlight a potential pitfall for users. It would have been nice to be able to highlight that comment so that it was more visible for anyone who was skimming the article and maybe didn't want to delve too deep into the conversation.

I think a single pinned post is fine for now (we also have a single pinned post in the feed, so hopefully some of that logic tracks).

**Comment by amywtlin** [comment]
Only being able to pin one comment sits right with me. I have updated the content of the issue to reflect it with the following addition:

-----
### Scope

We will focus on enabling authors and super admins & cross-site admins* to pin a comment, no matter their motivations behind doing such action. Hence we are opting for a ubiquitous term to describe this neutral action as "pinning a comment."

-----

*cross-site admins is a relative term, comparing to single-source/-channel/-page admins, such as a tag moderator who can only moderate authorized tags. (Open to suggestions on alternative ways to call this - I just wanted one term with clarity of what I meant)

**Comment by amywtlin** [comment]
Also, on this: 

> Who should be able to mark comments as best answers?

I think post authors, cross-site admins (which is referred to as [Admins](https://docs.google.com/document/d/1_MUj3IsbK1g5Cf2NyUk_MAoz0SIYzROLQ8XNdsCPfrE/edit) in our system right now) and super admins.

What do you think?

@ludwiczakpawel Can you help update the terms from best answer to pinning a comment throughout the issue? I don't want to break your things 🙈

**Comment by ludwiczakpawel** [comment]
@amywtlin you mean mockups? sure thing, i'll do it 👍

**Comment by amywtlin** [comment]
@ludwiczakpawel Yes please! I also just updated the issue :D

**Comment by aitchiss** [comment]
I've unassigned myself from this one, as I'm going to be floating into admin experience pod for a bit and not sure yet I've got the capacity to pick this up (I hadn't started any work on it yet)

**Comment by ludwiczakpawel** [comment]
@amywtlin Design has been updated ✅

**Comment by mikeydorje** [comment]
Closing due to inactivity — Please open a new issue with updated context.
